src/vs/editor/contrib/colorPicker/browser/color.ts

```diff
@@ -13,83 +13,137 @@ import { IModelService } from 'vs/editor/common/services/model';
+		colors.push(...documentColors);
+	}
+	return Array.isArray(documentColors);
 }
 
-export function getColorPresentations(model: ITextModel, colorInfo: IColorInformation, provider: DocumentColorProvider, token: CancellationToken): Promise<IColorPresentation[] | null | undefined> {
-	return Promise.resolve(provider.provideColorPresentations(model, colorInfo, token));
+function _getColorInfo(color: number[], range: IRange): IColorInformation {
+	const [red, green, blue, alpha] = color;
+	return {
+		range: range,
+		color: { red, green, blue, alpha }
+	};
 }
 
-CommandsRegistry.registerCommand('_executeDocumentColorProvider', function (accessor, ...args) {
+async function _computeColorData(options: ComputeColorPresentationsOptions | ComputeDocumentColorsOptions, provider: DocumentColorProvider, colorData: IColorType[], model: ITextModel, token: CancellationToken) {
+	if (options instanceof ComputeDocumentColorsOptions) {
+		return await _computeDocumentColors(provider, colorData, options.source, model, token);
+	} else {
+		return await _computeColorPresentations(provider, colorData, model, _getColorInfo(options.color, options.range));
+	}
+}
 
-	const [resource] = args;
-	if (!(resource instanceof URI)) {
-		throw illegalArgument();
+async function _findColorData<T extends ComputeDocumentColorsOptions | ComputeColorPresentationsOptions>(options: T, colorProviderRegistry: LanguageFeatureRegistry<DocumentColorProvider>, model: ITextModel, token: CancellationToken, isDefaultColorDecoratorsEnabled: boolean): Promise<T extends ComputeDocumentColorsOptions ? IExtColorData[] | IColorData[] : IColorPresentation[]>;
+async function _findColorData(options: ComputeDocumentColorsOptions | ComputeColorPresentationsOptions, colorProviderRegistry: LanguageFeatureRegistry<DocumentColorProvider>, model: ITextModel, token: CancellationToken, isDefaultColorDecoratorsEnabled: boolean): Promise<IExtColorData[] | IColorData[] | IColorPresentation[]> {
+	let validDocumentColorProviderFound = false;
+	let defaultDocumentColorProvider: DefaultDocumentColorProvider | undefined;
+	const colorData: IColorData[] | IExtColorData[] | IColorPresentation[] = [];
+	const documentColorProviders = colorProviderRegistry.ordered(model);
+	for (let i = documentColorProviders.length - 1; i >= 0; i--) {
+		const provider = documentColorProviders[i];
+		if (provider instanceof DefaultDocumentColorProvider) {
+			defaultDocumentColorProvider = provider;
+		} else {
+			try {
+				validDocumentColorProviderFound ||= await _computeColorData(options, provider, colorData, model, token);
+			} catch (e) {
+				onUnexpectedExternalError(e);
+			}
+		}
 	}
+	if (validDocumentColorProviderFound) {
+		return colorData;
+	}
+	if (!defaultDocumentColorProvider || !isDefaultColorDecoratorsEnabled) {
+		return [];
+	} else {
+		validDocumentColorProviderFound ||= await _computeColorData(options, defaultDocumentColorProvider, colorData, model, token);
+		return colorData;
+	}
```

Comment:
just a simpification..
```suggestion
  if (defaultProvider && isDefaultColorDecoratorsEnabled) {
    await _computeColorData(options, defaultProvider, colorData, model, token);
    return colorData;
  }
  return [];
```